activerecord/test/cases/adapter_test.rb

```diff
@@ -737,6 +737,15 @@ def teardown
         assert_predicate @connection, :active?
       end
 
+      test "statement-cached association queries are retried and result in a reconnect" do
+        p = Post.first
+
+        remote_disconnect @connection
```

Comment:
An idea I had to make these tests avoid the `remote_disconnect` dance was we could add `allow_retry`/`retryable` to `sql.active_record`. Then we could make very simple positive/negative assertions for each kind of query

Reply:
It would probably be better indeed.

Reply:
Making sure we're on the same page: do you think that's worth including in this PR or should it be separate?

Reply:
Could be an independent PR, and once merge you can complete this one.

Reply:
Rebased on top of #54454